internal/infer/infer.go

```diff
@@ -65,17 +65,25 @@ func Scope(
 // Type attempts to infer the data model from the data, possibly evaluating expressions.
 func Type(
 	data any,
-	internalDataModel *schema.ScopeSchema,
+	internalDataModel schema.Scope,
 	functions map[string]schema.Function,
 	workflowContext map[string][]byte,
 ) (schema.Type, error) {
-	if expression, ok := data.(expressions.Expression); ok {
-		expressionType, err := expression.Type(internalDataModel, functions, workflowContext)
+	switch expr := data.(type) {
+	case expressions.Expression:
+		expressionType, err := expr.Type(internalDataModel, functions, workflowContext)
 		if err != nil {
-			return nil, fmt.Errorf("failed to evaluate type of expression %s (%w)", expression.String(), err)
+			return nil, fmt.Errorf("failed to evaluate type of expression %s (%w)", expr.String(), err)
 		}
 		return expressionType, nil
+	case *OneOfExpression:
+		oneOfType, err := expr.Type(internalDataModel, functions, workflowContext)
+		if err != nil {
+			return nil, fmt.Errorf("failed to evaluate type of expression %s (%w)", expr.String(), err)
+		}
+		return oneOfType, nil
```

Comment:
I don't see the need for this duplicated code. It seems to be completely identical to the `expressions.Expression` case -- can they be combined?

Reply:
> It seems to be completely identical to the expressions.Expression case

That's because `expr` has different types in the two cases which implement the same functions.

@jaredoconnell, I would think that you could declare an interface which provides just the `Type()` and `String()` functions (which is all which are used here) and then use that interface as the `case` value for a common case:  inside the implementations of `Type()` and `String()` the code will know which type it was invoked on (that is, they will _not_ have common implementations -- rather, the two types will each implement the common _interface_).  So, I don't _think_ any differentiation would be required.

Reply:
I started implementing that option, but I believe that will just make the code more obscure. It just adds another type to learn what it does. There are only two instances where this is duplicated.

Reply:
In this case, if you add another type, it will fail when it hits the switch, whereas if we make it generic, it will "just work". The risk in this case is that the behavior may need to be different. It's a trade-off, and it may be worth it in a similar situation where the code is duplicated more.

Reply:
I'm not sure what you tried.  Here's a [toy example](https://go.dev/play/p/RPXl8lSuJ7_E) which behaves as I expected, which is based just on using an `interface` without having to resort to a _generic_ function.  (I agree that using a generic function would be an option, but it strikes me that using an interface is a more appropriate approach, here -- among other things, it makes the type-dependency inherent, so that we don't have to use an explicit type check inside the function body or rely on the function instantiation to discriminate if it needs to do something type-specific.)

As for the amount of duplication, _any_ duplication is a possibility for bugs as the code evolves; so, I would avoid duplication unless avoiding it requires substantially more code/complication/etc. than keeping it.  (And, I don't think a single interface type declaration represents a significant amount of code.)

Reply:
That example matches the method that I used that I wrote and then discarded.

Reply:
And you discarded it because it fails when it hits the switch?  How did your code fail when mine seems to succeed?

Reply:
It worked. I just did not like it because it made it more obscure.

Reply:
Ah, OK.

Then I guess I was completely misinterpreting your earlier statement,
> In this case, if you add another type, it will fail when it hits the switch, whereas if we make it generic, it will "just work".

So, when you said "if you add another type" you weren't referring to the immediate addition of an interface type that I was proposing but rather to a hypothetical future evolution of the code which would potentially need another branch in the `switch` statement?  (And, when you said "make it generic", you weren't actually talking about using a Go generic function, right?)

If the hypothetical "new type" implements the interface which I'm proposing for the expressions, then the `switch` statement would not need to be changed, and things would indeed "just work"...provided that the new type's implementation of `Type()` and `String()` were appropriate.  And, this is a _feature_ of using an interface in this way (which is another part of the value which offsets the cost of the obscurity that so concerns you.)

Reply:
That is correct. It may be worth a refactor in the future if more types are introduced.